ros/src/computing/perception/localization/packages/ekf_localizer/src/ekf_localizer.cpp

```diff
@@ -0,0 +1,672 @@
+    start = std::chrono::system_clock::now();
+    measurementUpdateTwist(*current_twist_ptr_);
+    elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::system_clock::now() - start).count();
+    DEBUG_INFO("[EKF] measurementUpdateTwist calculation time = %f [ms]", elapsed * 1.0e-6);
+    DEBUG_INFO("------------------------- end twist -------------------------\n");
+  }
+
+  /* set current pose, twist */
+  setCurrentResult();
+
+  publishEstimatedPose();
+}
+
+void EKFLocalizer::showCurrentX()
+{
+  if (show_debug_info_)
+  {
+    Eigen::MatrixXd X(dim_x_, 1);
+    ekf_.getLatestX(X);
+    DEBUG_PRINT_MAT(X.transpose());
+  }
+}
+
+/*
+ * setCurrentResult
+ */
+void EKFLocalizer::setCurrentResult()
+{
+  current_ekf_pose_.header.frame_id = pose_frame_id_;
+  current_ekf_pose_.header.stamp = ros::Time::now();
+  current_ekf_pose_.pose.position.x = ekf_.getXelement(IDX::X);
+  current_ekf_pose_.pose.position.y = ekf_.getXelement(IDX::Y);
+
+  tf2::Quaternion q_tf;
+  double roll, pitch, yaw;
+  if (current_pose_ptr_ != nullptr)
+  {
+    current_ekf_pose_.pose.position.z = current_pose_ptr_->pose.position.z;
+    tf2::fromMsg(current_pose_ptr_->pose.orientation, q_tf); /* use Pose pitch and roll */
+    tf2::Matrix3x3(q_tf).getRPY(roll, pitch, yaw);
+  }
+  else
+  {
+    current_ekf_pose_.pose.position.z = 0.0;
+    roll = 0;
+    pitch = 0;
+  }
+  yaw = ekf_.getXelement(IDX::YAW) + ekf_.getXelement(IDX::YAWB);
+  q_tf.setRPY(roll, pitch, yaw);
+  tf2::convert(q_tf, current_ekf_pose_.pose.orientation);
+
+  current_ekf_twist_.header.frame_id = "base_link";
+  current_ekf_twist_.header.stamp = ros::Time::now();
+  current_ekf_twist_.twist.linear.x = ekf_.getXelement(IDX::VX);
+  current_ekf_twist_.twist.angular.z = ekf_.getXelement(IDX::WZ);
+}
+
+/*
+ * timerTFCallback
+ */
+void EKFLocalizer::timerTFCallback(const ros::TimerEvent &e)
+{
+  if (current_ekf_pose_.header.frame_id == "")
+    return;
+
+  geometry_msgs::TransformStamped transformStamped;
+  transformStamped.header.stamp = ros::Time::now();
+  transformStamped.header.frame_id = current_ekf_pose_.header.frame_id;
+  transformStamped.child_frame_id = "ekf_pose";
+  transformStamped.transform.translation.x = current_ekf_pose_.pose.position.x;
+  transformStamped.transform.translation.y = current_ekf_pose_.pose.position.y;
+  transformStamped.transform.translation.z = current_ekf_pose_.pose.position.z;
+
+  transformStamped.transform.rotation.x = current_ekf_pose_.pose.orientation.x;
+  transformStamped.transform.rotation.y = current_ekf_pose_.pose.orientation.y;
+  transformStamped.transform.rotation.z = current_ekf_pose_.pose.orientation.z;
+  transformStamped.transform.rotation.w = current_ekf_pose_.pose.orientation.w;
+
+  tf_br_.sendTransform(transformStamped);
+}
+
+/*
+ * getTransformFromTF
+ */
+bool EKFLocalizer::getTransformFromTF(std::string parent_frame, std::string child_frame, geometry_msgs::TransformStamped &transform)
```

Comment:
For instance, tests for these methods.

I see you have a timer and you repeat 10 times in case of the TF cannot be obtained.

1. Why 10 times?
1. What happens if TF cannot be obtained after 10 tries?

Reply:
any comment on this?

Reply:
I'm taking codes for this, please wait.

Reply:
Sorry, for your questions,

1. -> no reason. I can change to any number  if you have a suggestion.
2. -> return false when TF cannot be obtained, then parent function check the return value. If it is false, print ROS_WARN(), then ingore action.

Reply:
I couldn't write a test of this funtion for some reasons (coundn't send transform while tf_listener alive which is born in this function).
For 1., I changed 10 times to 50 times & also changed sleep time from 0.5 to 0.1. (for more fast reaction)

Reply:
what would happen if you can't get the Tf after `N` times?
* Localization will fail?
* will the EKF stop predicting?